### ISSUE TITLE
Variables that aren't necessary for an operation shouldn't be sent

### ISSUE THREAD

**Issue by sgrove** [graphql/graphiql#981]
Right now GraphiQL will send the full value of the `variables` form, even if they query itself doesn't make use of them. This can be pretty cumbersome on e.g. Hasura where it'll refuse to run a query if it sees unexpected variables present, but you might have multiple operations with their own variables and you want to be able to run them in quick succession.

It'd be nice if, just like auto-setting an `operationName`, GraphiQL plucked only the relevant variables to be sent to the server.

**Comment by benjie** [comment]
I worry that this is too much in the realms of "magic." If the user enters variables into the variable editor and they are not sent to the server I think that could be very confusing (and misleading when they copy those same variables into their application and then it doesn't work even though it worked in GraphiQL). As such, I think this would make sense for an opt-in plugin but should not be part of the core implementation.

**Comment by Neitsch** [comment]
@sgrove - thank you for the suggestion. I agree with Benjie, that at this time it is probably too much added magic. I believe that the spec does allow you to pass additional (unused) variables in the request. Would it maybe be possible to raise this issue with Hasura? :)

**Comment by acao** [comment]
Agreed this would be great for a plugin, not what we'd expect as spec behavior

**Comment by acao** [comment]
I've even seen these expected failures as part of application logic. For example, this one case where someone was firing off a mutation with a very complex, nested tree of objects. Instead of duck typing the payload in the client, they would attempt the mutation over the wire, and use the server to validate and then use the error message payload to show validation errors to the user.  I'm sure there are many implementations out there in the wild that depend on being able to send potentially invalid variables payloads for server side validation.
